### PR TITLE
LoadColumnName support in TextLoader and auto map class members 

### LINKED ISSUE #4480: LoadColumn should not be required when loading from a file with headers
### System information

- **OS version/distro**: Windows 10 (10.0.18362)
- **.NET Version (eg., dotnet --info)**:  3.0.100

### Issue

- **What did you do?**
Created a type anottated with `LoadColumnName`
Created a `TextLoader` from this type and `useHeaders` set to `true` and called the `Load` method

- **What happened?**
System.InvalidOperationException: 'Property 'IESCode' is missing the LoadColumnAttribute attribute'

- **What did you expect?**
The Load method should map the properties from the file headers when available, not forcing the use of column index.

### Source code / logs

```fsharp
System.InvalidOperationException: 'Property 'IESCode' is missing the LoadColumnAttribute attribute'
```

```fsharp
type EnadeMicrodata =
    {
        [<LoadColumnName("CO_IES")>]
        IESCode: int

        [<LoadColumnName("CO_CATEGAD")>]
        AdministrativeCategory: int
    }

let mlContext = MLContext ()
let _textLoader = mlContext.Data.CreateTextLoader<EnadeMicrodata>(hasHeader = true).Load("MICRODADOS_ENADE_2017.txt")
```

### PR DESCRIPTION
**Pull request by kevencarneiro**
This pull request resolves #4480.

These features were add in this pull request:
* Support of `LoadColumnName` attribute in `TextLoader` class (when loaded file has headers).
* Possibility to ignore a member when using `CreateTextLoader` using the `LoadColumnIgnoreAttribute`
* Feature to map the `TInput` properties/fields name to column of TextLoader (when loaded file has headers)

### CODE REVIEW

**Comment by kevencarneiro** [comment]
Some tests are failing due to changes in `LoadFromTextFile` method that now has the input passed as sample data and including SlotNames in files. Shouldn't this verification looks if every existing property is present, but allow new properties to be inserted?
The `EntryPointCatalog` is also failing, but the verification shouldn't follow the same logic? Verify if all existing EntryPoints are present, but allow new EntryPoints?

Or in this case is OK to replace the baseline test file?

**Comment by harishsk** [comment]
@kevencarneiro Thank you for contribution. 
I will review your proposed changes and get back to you soon.

File: Microsoft.ML.sln [hunk 1]
```diff
@@ -94,7 +94,7 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.ML.CpuMath.Perfor
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.ML.CpuMath.UnitTests", "test\Microsoft.ML.CpuMath.UnitTests\Microsoft.ML.CpuMath.UnitTests.csproj", "{E97D8B5A-3035-4D41-9B0D-77FF8FB8D132}"
 EndProject
-Project("{6EC3EE1D-3C4E-46DD-8F32-0CC8E7565705}") = "Microsoft.ML.FSharp.Tests", "test\Microsoft.ML.FSharp.Tests\Microsoft.ML.FSharp.Tests.fsproj", "{802233D6-8CC0-46AD-9F23-FEE1E9AED9B3}"
+Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "Microsoft.ML.FSharp.Tests", "test\Microsoft.ML.FSharp.Tests\Microsoft.ML.FSharp.Tests.fsproj", "{802233D6-8CC0-46AD-9F23-FEE1E9AED9B3}"
 EndProject
```
Comment (harishsk) [inline Microsoft.ML.sln:98]: Any idea why this change is necessary?

Comment (kevencarneiro) [reply Microsoft.ML.sln:98]: This change was made automatically when I opened the project in Visual Studio for Mac.

**Comment by harishsk** [comment]
The failures you are seeing with respect to slotnames is odd. I am looking into it and will post an update soon.

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs [hunk 2]
```diff
@@ -194,25 +217,27 @@ public static class TextLoaderSaverCatalog
...
                 throw Contracts.ExceptParam(nameof(path), "File does not exist at path: {0}", path);
             }
 
-            // REVIEW: it is almost always a mistake to have a 'trainable' text loader here.
-            // Therefore, we are going to disallow data sample.
+            var source = new MultiFileSource(path);
             return TextLoader.CreateTextLoader<TInput>(CatalogUtils.GetEnvironment(catalog), hasHeader, separatorChar,
-                allowQuoting, allowSparse, trimWhitespace).Load(new MultiFileSource(path));
+                allowQuoting, allowSparse, trimWhitespace, autoMapLoadColumns, dataSample: source).Load(source);
         }
```
Comment (harishsk) [inline src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs:241]: The onnx related test failures you are seeing are a result of this line. 
When you supply a data sample and the file has a header, the header fields are added as annotations to columns which get saved in the onnx model when the pipeline is exported. Since the earlier models were created without the data sample, they don't have annotations and you are seeing failures when comparing against the baseline. 

It is debatable whether we should always add the column annotations when a header is present. There is a comment in the older version that explicitly disallows a data sample. This was prior to my time, so I am not sure what the background is on this. 

@eerhardt / @yaeldekel : Do either of you know why it is considered a mistake to have a 'trainable' text loader here and using a data sample?

@kevencarneiro : It is possible to take your PR and retain the old behavior if you change the lines above to the following:
```
            var source = new MultiFileSource(path);
            var dataSample = autoMapLoadColumns ? source : null;
            return TextLoader.CreateTextLoader<TInput>(CatalogUtils.GetEnvironment(catalog), hasHeader, separatorChar,
                allowQuoting, allowSparse, trimWhitespace, autoMapLoadColumns, dataSample: dataSample).Load(source);
```

**Comment by harishsk** [comment]
@eerhardt / @yaeldekel 
Can at least one of you please review this PR in addition to myself?

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs [hunk 3]
```diff
@@ -1479,16 +1551,38 @@ void ICanSaveModel.Save(ModelSaveContext ctx)
...
+
+                var mappingLoadColumnAttr = memberInfo.GetCustomAttribute<LoadColumnAttribute>();
+                var mappingLoadColumnNameAttr = memberInfo.GetCustomAttribute<LoadColumnNameAttribute>();
+                var memberInfoType = memberInfo is FieldInfo ? "Field" : "Property";
+
+                if (!hasHeader && mappingLoadColumnAttr == null)
+                    throw host.Except(
+                        $"You should set {nameof(hasHeader)} to true in order to use {nameof(LoadColumnNameAttribute)}. " +
+                        $"To automatically map the name of {memberInfoType} to {nameof(LoadColumnNameAttribute)}, set {nameof(autoMapLoadColumns)} to true. " +
+                        $"To specify a fixed column index, use {nameof(LoadColumnAttribute)}. " +
```
Comment (harishsk) [inline src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs:1567]: If autoMalLoadColumns is true, then the hasHeader must be true. 
Can we factor this out of the loop as an explicit parameter check at the beginning of the function?

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs [hunk 4]
```diff
@@ -1479,16 +1551,38 @@ void ICanSaveModel.Save(ModelSaveContext ctx)
...
+                        $"Invalid {memberInfoType} name: {memberInfo.Name}.");
```
Comment (harishsk) [inline src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs:1568]: This should go in the documentation. Can you please shorten the exception messages (here and below) to be focused on the specific exception?

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs [hunk 5]
```diff
@@ -1479,16 +1551,38 @@ void ICanSaveModel.Save(ModelSaveContext ctx)
...
+                        $"To ignore this member, use {nameof(LoadColumnIgnoreAttribute)}. " +
+                        $"Invalid {memberInfoType} name: {memberInfo.Name}.");
 
                 var mappingAttrName = memberInfo.GetCustomAttribute<ColumnNameAttribute>();
 
                 var column = new Column();
                 column.Name = mappingAttrName?.Name ?? memberInfo.Name;
-                column.Source = mappingAttr.Sources.ToArray();
+                column.Source =  (mappingLoadColumnNameAttr != null)
+                    ? mappingLoadColumnNameAttr.Sources.Select(x => new Range(x)).ToArray()
```
Comment (harishsk) [inline src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs:1582]: What happens if two or more of LoadColumn, LoadColumnName and ColumnName are specified and there is a conflict? Can you please identify that and throw an exception.

**Comment by harishsk** [comment]
The failure in EntryPointCatalog is because a new parameter has been added to the public API and it is failing a comparison against the current baseline. The current baseline is stored in test\BaselineOutput\Common\EntryPoints\core_manifest.json
I think it is okay to update the baseline in this case. @yaeldekel : Can you please verify?

---
In reply to: [614830681](https://github.com/dotnet/machinelearning/pull/5021#issuecomment-614830681) [](ancestors = 614830681)

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs (hunk 2 above)
Comment (yaeldekel) [reply src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs:241]: Regarding the reason not to have a trainable text loader in the `LoadFromTextFile` API: When you create a trainable text loader, some parameters of the text loader are learned from the data sample. Specifically, slot names, if they exist, and in some cases the cardinality of some columns in the data (the range can be specified with only a minimum and no maximum, indicating the text loader needs to figure out the end of the range on its own). Lets imagine that a user loads a data view using `LoadFromTextFile`, and fits a pipeline using the data view to create a model. They then load new data using `LoadFromTextFile`, and try to apply the model to it. If the data view returned from `LoadFromTextFile` uses a trainable text loader, it could end up having different schemas depending on the actual data in the text file, which might cause the model to fail when applied to the second dataset. 



---
In reply to: [409767689](https://github.com/dotnet/machinelearning/pull/5021#discussion_r409767689) [](ancestors = 409767689)

Comment (yaeldekel) [reply src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs:241]: Given this, I think that the new `autoMapLoadColumns` option should not be present in `LoadFromTextFile`, only in the `CreateTextLoader` APIs.

---
In reply to: [411977600](https://github.com/dotnet/machinelearning/pull/5021#discussion_r411977600) [](ancestors = 411977600,409767689)

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs [hunk 6]
```diff
@@ -85,6 +85,16 @@ public static class TextLoaderSaverCatalog
             IMultiStreamSource dataSample = null)
             => new TextLoader(CatalogUtils.GetEnvironment(catalog), options, dataSample);
 
+        // 1.4.0 BACKCOMPAT OVERLOAD -- DO NOT TOUCH
```
Comment (yaeldekel) [inline src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs:88]: >1.4.0 BACKCOMPAT OVERLOAD -- DO NOT TOUCH [](start = 11, length = 41)

It seems to me that these overloads are not needed, for two reasons:
First of all, since none of the arguments here are optional, it doesn't actually provide backwards compatibility (if a user has code that only uses some of the arguments, it will not compile with this overload). Second of all, since the arguments in the current signature **are** optional, adding another optional argument at the end of the argument list, wouldn't break existing code, so it is backwards compatible. 
Is there a different reason to have these overloads that I am not aware of?

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs [hunk 7]
```diff
@@ -1479,16 +1551,38 @@ void ICanSaveModel.Save(ModelSaveContext ctx)
...
-                    throw host.Except($"{(memberInfo is FieldInfo ? "Field" : "Property")} '{memberInfo.Name}' is missing the {nameof(LoadColumnAttribute)} attribute");
+                var ignoreLoadColumnAttr = memberInfo.GetCustomAttribute<LoadColumnIgnoreAttribute>();
+                if (ignoreLoadColumnAttr != null && ignoreLoadColumnAttr.Ignore)
+                    continue;
+
+                var mappingLoadColumnAttr = memberInfo.GetCustomAttribute<LoadColumnAttribute>();
+                var mappingLoadColumnNameAttr = memberInfo.GetCustomAttribute<LoadColumnNameAttribute>();
+                var memberInfoType = memberInfo is FieldInfo ? "Field" : "Property";
+
+                if (!hasHeader && mappingLoadColumnAttr == null)
```
Comment (yaeldekel) [inline src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs:1563]: >mappingLoadColumnAttr [](start = 34, length = 21)

It seems that you are assuming here that if `mappingLoadColumnAttr == null`, then necessarily `mappingLoadColumnNameAttr != null`, but you are not checking to make sure this is true.

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs (hunk 3 above)
Comment (yaeldekel) [reply src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs:1567]: I think that other than to be able to factor this check out of the loop, the `autoMapLoadColumns` argument is not needed, so I would prefer not to have it in the signature of the method at all.

---
In reply to: [409848508](https://github.com/dotnet/machinelearning/pull/5021#discussion_r409848508) [](ancestors = 409848508)

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs [hunk 8]
```diff
@@ -1479,16 +1551,38 @@ void ICanSaveModel.Save(ModelSaveContext ctx)
             for (int index = 0; index < memberInfos.Length; index++)
             {
                 var memberInfo = memberInfos[index];
-                var mappingAttr = memberInfo.GetCustomAttribute<LoadColumnAttribute>();
 
-                if (mappingAttr == null)
-                    throw host.Except($"{(memberInfo is FieldInfo ? "Field" : "Property")} '{memberInfo.Name}' is missing the {nameof(LoadColumnAttribute)} attribute");
+                var ignoreLoadColumnAttr = memberInfo.GetCustomAttribute<LoadColumnIgnoreAttribute>();
```
Comment (yaeldekel) [inline src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs:1555]: >LoadColumnIgnoreAttribute [](start = 73, length = 25)

Why is this being added? What is the scenario where you would need to have a public field/property in the class but not have it loaded by the text loader?
If this is really necessary, then I would instead of introducing this new class, have the use the `LoadColumnNameAttribute`, and have an empty or null `Name` indicate that this field should be ignored. There is also a `NoColumnAttribute` class that can alternatively be used.

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs [hunk 9]
```diff
@@ -700,6 +734,33 @@ public Bindings(TextLoader parent, Column[] cols, IMultiStreamSource headerFile,
...
+
+                        var header = Parser.GetHeaderFields(parent, lines[0]);
+                        var headerMapping = Enumerable.Zip(header.Indices, header.Spans,
+                            (index, span) => new
+                            {
+                                Index = index,
+                                Span = span
+                            })
+                            .Where(x => !x.Span.IsEmpty)
+                            .GroupBy(x => x.Span.ToString().ToLowerInvariant());
```
Comment (yaeldekel) [inline src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs:750]: >ToLowerInvariant [](start = 60, length = 16)

Column names are usually case sensitive, so there is no need to call `ToLowerInvariant`.

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs [hunk 10]
```diff
@@ -495,6 +518,13 @@ public class Options
                 HelpText = "Data file has header with feature names. Header is read only if options 'hs' and 'hf' are not specified.")]
             public bool HasHeader;
 
+            /// <summary>
+            /// Whether the members name of the class passed as type argument should be mapped to <see cref="LoadColumnNameAttribute"/>.
+            /// When <see langword="true"/> requires the <see cref="HasHeader"/> option to be <see langword="true"/>.
+            /// </summary>
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Use class properties as LoadColumnNameAttribute to read from text file", ShortName = "automap")]
+            public bool AutoMapLoadColumns;
```
Comment (yaeldekel) [inline src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs:526]: >AutoMapLoadColumns [](start = 24, length = 18)

This is only used in the generic overload, that creates a `TextLoader` from a given type, and not in the overload that takes an `Options` object, so it doesn't need to be a part of the `Options` class.

**Comment by yaeldekel** [comment]
Hi @kevencarneiro, thanks for tackling this issue!

As @harishsk mentioned, the failure in the `EntryPointCatalog` test is being caused because the `Options` class has changed, and it is used as the input to the `TextLoader`'s entry point. But as I commented in the TextLoader.cs file, I think the `Options` class does not need to change, so it will not be necessary to change the entrypoint catalog baseline. 

The other failure will also be solved if you revert the change that adds the data sample to the constructor of `TextLoader` in the `LoadFromTextFile` API.

---
In reply to: [613231161](https://github.com/dotnet/machinelearning/pull/5021#issuecomment-613231161) [](ancestors = 613231161)

File: Microsoft.ML.sln (hunk 1 above)
Comment (harishsk) [reply Microsoft.ML.sln:98]: Can you please revert this change and verify if everything still works?

**Comment by michaelgsharp** [comment]
Closing for now due to inactivity. If you would like to resume working on this please feel free to either reopen this PR or create a new one.
